parlai/agents/bart/bart.py

```diff
@@ -0,0 +1,181 @@
+        Download and convert BART pre-trained models.
+
+        Additionally, convert `init-fairseq-model` if necessary.
+
+        :param opt:
+            ParlAI-parsed options
+
+        :return opt:
+            return opt with BART-specific args.
+        """
+        if not opt.get('converting'):
+            download(opt['datapath'])
+            opt['init_model'] = os.path.join(
+                opt['datapath'], 'models/bart/bart_large/model'
+            )
+        if opt.get('init_fairseq_model'):
+            opt = self._convert_model(opt)
+        opt.update(BART_ARGS)
+        compare_init_model_opts(opt, opt)
+        return opt
+
+    def _convert_model(self, opt: Opt) -> Opt:
+        """
+        Convert fairseq init model to ParlAI Model.
+
+        :param opt:
+            options
+
+        :return opt:
+            return opt with new init_model path
+        """
+        model_name = os.path.split(opt['init_fairseq_model'])[-1]
+        args = CONVERSION_ARGS.copy()
+
+        args['input'] = [opt['init_fairseq_model']]
+        if opt.get('model_file') and not os.path.exists(opt['model_file']):
+            args['output'] = opt['model_file']
+        else:
+            args['output'] = os.path.join(
+                opt['datapath'], 'models/converted_fairseq_models/', model_name
+            )
+
+        ConversionScript.main(**args)
+        opt['init_model'] = args['output']
+        return opt
+
+    def build_model(self) -> BartModel:
+        """
+        Build and return model.
+        """
+        model = BartModel(self.opt, self.dict)
+        if self.opt['embedding_type'] != 'random':
+            self._copy_embeddings(
+                model.encoder.embeddings.weight, self.opt['embedding_type']
+            )
+        return model
+
+    def _set_text_vec(
+        self, obs: Message, history: History, truncate: Optional[int]
+    ) -> Message:
+        """
+        Override to prepend start token and append end token.
+        """
+        obs = super()._set_text_vec(obs, history, truncate)
+        if 'text' not in obs or 'text_vec' not in obs:
+            return obs
+        vec = obs['text_vec']
+        if truncate is not None:
+            vec = torch.LongTensor(
+                self._check_truncate(obs['text_vec'], truncate - 2, True)
+            )
+        obs.force_set(
+            'text_vec',
+            torch.cat(
+                [
+                    vec.new_tensor([self.dict[self.dict.start_token]]),
+                    vec,
+                    vec.new_tensor([self.dict[self.dict.end_token]]),
+                ],
+                0,
+            ),
+        )
+        return obs
+
+    def _get_initial_decoder_input(self, bsz: int, beam_size: int, dev: str):
+        """
+        Override to seed decoder with EOS token.
+
+        See docstring for `BartAgent._generate` for more details.
+        """
+        return torch.LongTensor([self.END_IDX]).expand(bsz * beam_size, 1).to(dev)
```

Comment:
hey maybe you already fixed this, but you might want to check that this doesn't cause problems with beam search, etc., since it will end once we see an end token. also see `_v2t`

Reply:
we call this function immediately before seeding the decoder, so beam search will only end if the decoder produces EOS as the next token